### PR TITLE
Synchronize OnDemand OS ConnectionManager propagation data

### PR DESCRIPTION
**Pull request by MBoldyrev**
### Description of the Change

This change makes OdOs Connection Manager aware of the round for which its destination peers are valid. Now it will wait for the peers data to update if the batch it gets has a newer round, or set the destination of the batch to an actual round if the requested destination is outdated.

### Benefits

Aimed at mitigation of (one of causes of) [the lost proposal bug|https://soramitsu.atlassian.net/browse/IR-1888].

### Possible Drawbacks 

Bugs!

### CODE REVIEW

File: irohad/ordering/impl/on_demand_connection_manager.cpp [hunk 1]
```diff
@@ -5,37 +5,73 @@
...
+  initializeConnections(inital_propagation_params);
 }
 
-void OnDemandConnectionManager::onBatches(consensus::Round round,
+void OnDemandConnectionManager::onBatches(const consensus::Round round,
                                           CollectionType batches) {
   std::shared_lock<std::shared_timed_mutex> lock(mutex_);
 
+  // check if we have appropriate propagation data:
+  if (current_round_ != round) {
```
Comment (l4l) [inline irohad/ordering/impl/on_demand_connection_manager.cpp]: might be moved to separate function and reworked with early return, e.g:
```cpp
void func(...) {
  if (current_round_ == round) {
    return
  }
  ...
}
```

File: irohad/ordering/impl/on_demand_connection_manager.cpp [hunk 2]
```diff
@@ -5,37 +5,52 @@
...
+  initializeConnections(inital_propagation_params);
 }
 
-void OnDemandConnectionManager::onBatches(consensus::Round round,
+void OnDemandConnectionManager::onBatches(const consensus::Round round,
                                           CollectionType batches) {
   std::shared_lock<std::shared_timed_mutex> lock(mutex_);
 
+  // check if we have an appropriate propagation data:
+  if (!verifyPropagtionData(round, lock)) {
```
Comment (l4l) [inline irohad/ordering/impl/on_demand_connection_manager.cpp]: It's hard to follow the function logic. I'd suggest passing waiting callback instead, e.g:

```cpp
verifyPropagationData(round, propagation_data_update_cv_.wait_for(
          lock, kPropagationDataWaitTimeout, [this, &request_round] {
            return not(this->current_round_ < request_round);
          }));
```

Comment (l4l) [reply irohad/ordering/impl/on_demand_connection_manager.cpp]: or maybe even remove function and inline it's body here

**Comment by MBoldyrev** [comment]
The issue of this PR is supposed to be solved after the base branch is merged to dev.
